File: push/src/main/java/com/xiaomi/push/service/XMPushServiceLifecycleDelegate.java
package com.xiaomi.push.service;

import android.content.Intent;
import com.xiaomi.channel.commonutils.android.SystemUtils;
import com.xiaomi.channel.commonutils.logger.MyLog;
import com.xiaomi.channel.commonutils.misc.BuildSettings;
import com.xiaomi.push.service.timers.Alarm;
import com.xiaomi.slim.SlimConnection;
import com.xiaomi.smack.ConnectionConfiguration;
import com.xiaomi.stats.StatsHandler;
import com.xiaomi.tinyData.TinyDataCacheProcessor;
import com.xiaomi.tinyData.TinyDataManager;

final class XMPushServiceLifecycleDelegate {
    private final XMPushService service;
    private final XMPushServiceLifecycleInfrastructure infrastructure;
    private final XMPushServiceLifecycleRuntime runtime;

    XMPushServiceLifecycleDelegate(XMPushService xMPushService) {
        this.service = xMPushService;
        this.infrastructure = new XMPushServiceLifecycleInfrastructure(xMPushService);
        this.runtime = new XMPushServiceLifecycleRuntime(xMPushService);
    }

    void onCreate() {
        MyLog.init(this.service.getApplicationContext());
        SystemUtils.initialize(this.service);
        MIPushAccount mIPushAccount = PushAccountRuntime.applyStoredAccountEnvironment(this.service, "XMPushService.onCreate");
        if (mIPushAccount != null) {
            BuildSettings.setEnvType(mIPushAccount.envType);
        }
        this.infrastructure.installMessenger();
        PushHostManagerFactory.init(this.service);
        ConnectionConfiguration connectionConfiguration = this.infrastructure.createConnectionConfiguration();
        connectionConfiguration.setDebuggerEnabled(true);
        this.service.setConnectionConfiguration(connectionConfiguration);
        SlimConnection slimConnection = new SlimConnection(this.service, connectionConfiguration);
        slimConnection.addConnectionListener(this.service);
        this.service.setSlimConnection(slimConnection);
        this.service.setClientEventDispatcher(this.service.createClientEventDispatcher());
        Alarm.initialize(this.service);
        this.service.setPacketSync(new PacketSync(this.service));
        this.service.setReconnectionManager(new ReconnectionManager(this.service));
        new CommonPacketExtensionProvider().register();
        StatsHandler.getInstance().init(this.service);
        this.service.setJobController(new JobScheduler("Connection Controller Thread"));
        this.runtime.configureClientChangeListener();
        if (this.service.canOpenForegroundService()) {
            this.service.enableForegroundService();
        }
        TinyDataManager.getInstance(this.service).addUploader(new LongConnUploader(this.service), TinyDataManager.UPLOADER_PUSH_CHANNEL);
        this.service.addPingCallBack(new TinyDataCacheProcessor(this.service));
        this.service.executeJob(new InitJob(this.service));
        this.service.addNetworkListener(Sync.getInstance(this.service));
        if (this.service.isPushEnabled()) {
            this.service.ensureConnectionChangeReceiver();
        }
        if (PushConstants.PUSH_SERVICE_PACKAGE_NAME.equals(this.service.getPackageName())) {
            this.infrastructure.installPowerModeObservers();
            this.infrastructure.installFalldownReceiver();
        }
        this.infrastructure.persistCreationLog(mIPushAccount);
    }

    void onDestroy() {
        ConnectionChangeReceiver connectionChangeReceiver = this.service.getConnectionChangeReceiver();
        if (connectionChangeReceiver != null) {
            this.service.unregisterReceiverSafely(connectionChangeReceiver);
            this.service.clearConnectionChangeReceiver();
        }
        ScreenStateReceiver screenStateReceiver = this.service.getScreenStateReceiver();
        if (screenStateReceiver != null) {
            this.service.unregisterReceiverSafely(screenStateReceiver);
            this.service.clearScreenStateReceiver();
        }
        this.infrastructure.unregisterPowerModeObservers();
        this.service.clearNetworkListeners();
        this.service.getJobController().removeAllJobs();
        this.service.executeJob(new XMPushService.Job(2) { // from class: com.xiaomi.push.service.XMPushServiceLifecycleDelegate.1
            @Override // com.xiaomi.push.service.XMPushService.Job
            public String getDesc() {
                return "disconnect for service destroy.";
            }

            @Override // com.xiaomi.push.service.XMPushService.Job
            public void process() {
                if (XMPushServiceLifecycleDelegate.this.service.getCurrentConnection() != null) {
                    XMPushServiceLifecycleDelegate.this.service.getCurrentConnection().disconnect(15, null);
                    XMPushServiceLifecycleDelegate.this.service.clearCurrentConnection();
                }
            }
        });
        this.service.executeJob(new KillJob(this.service));
        PushClientsManager.getInstance().removeAllClientChangeListeners();
        PushClientsManager.getInstance().resetAllClients(this.service, 15);
        PushClientsManager.getInstance().removeActiveClients();
        this.service.getSlimConnection().removeConnectionListener(this.service);
        ServiceConfig.getInstance().clear();
        Alarm.stop();
        this.service.clearPingCallbacks();
    }

    void onStart(Intent intent, int i) {
        long currentTimeMillis = System.currentTimeMillis();
        if (intent == null) {
            MyLog.e("onStart() with intent NULL");
        } else {
            MyLog.w(String.format("onStart() with intent.Action = %s, chid = %s, pkg = %s|%s", intent.getAction(), intent.getStringExtra(PushConstants.EXTRA_CHANNEL_ID), intent.getStringExtra(PushConstants.EXTRA_PACKAGE_NAME), intent.getStringExtra(PushConstants.MIPUSH_EXTRA_APP_PACKAGE)));
        }
        if (intent != null && intent.getAction() != null) {
            if (PushServiceConstants.ACTION_TIMER.equalsIgnoreCase(intent.getAction()) || PushServiceConstants.ACTION_CHECK_ALIVE.equalsIgnoreCase(intent.getAction())) {
                if (this.service.getJobController().isBlocked()) {
                    MyLog.e("ERROR, the job controller is blocked.");
                    PushClientsManager.getInstance().resetAllClients(this.service, 14);
                    this.service.stopSelf();
                } else {
                    this.service.executeJob(new IntentJob(this.service, intent));
                }
            } else if (!PushServiceConstants.ACTION_NETWORK_STATUS_CHANGED.equalsIgnoreCase(intent.getAction())) {
                this.service.executeJob(new IntentJob(this.service, intent));
            }
        }
        long currentTimeMillis2 = System.currentTimeMillis() - currentTimeMillis;
        if (currentTimeMillis2 > 50) {
            MyLog.v("[Prefs] spend " + currentTimeMillis2 + " ms, too more times.");
        }
    }

    int onStartCommand(Intent intent, int i, int i2) {
        onStart(intent, i2);
        return 1;
    }

    void networkChanged() {
        this.runtime.networkChanged();
    }

    void connectionClosed(com.xiaomi.smack.Connection connection, int i, Exception exc) {
        this.runtime.connectionClosed(connection, i, exc);
    }

    void connectionStarted(com.xiaomi.smack.Connection connection) {
        this.runtime.connectionStarted(connection);
    }

    void postOnCreate() {
        this.runtime.postOnCreate();
    }

    void reconnectionFailed(com.xiaomi.smack.Connection connection, Exception exc) {
        this.runtime.reconnectionFailed(connection, exc);
    }

    void reconnectionSuccessful(com.xiaomi.smack.Connection connection) {
        this.runtime.reconnectionSuccessful(connection);
    }
}

File: push/src/main/java/com/xiaomi/push/service/XMPushServiceLifecycleDelegate.kt
package com.xiaomi.push.service

import android.content.Intent
import com.xiaomi.channel.commonutils.android.SystemUtils
import com.xiaomi.channel.commonutils.logger.MyLog
import com.xiaomi.channel.commonutils.misc.BuildSettings
import com.xiaomi.push.service.timers.Alarm
import com.xiaomi.slim.SlimConnection
import com.xiaomi.smack.ConnectionConfiguration
import com.xiaomi.stats.StatsHandler
import com.xiaomi.tinyData.TinyDataCacheProcessor
import com.xiaomi.tinyData.TinyDataManager

class XMPushServiceLifecycleDelegate(
    private val service: XMPushService,
) {
    private val infrastructure = XMPushServiceLifecycleInfrastructure(service)
    private val runtime = XMPushServiceLifecycleRuntime(service)

    fun onCreate() {
        MyLog.init(service.applicationContext)
        SystemUtils.initialize(service)
        val account = PushAccountRuntime.applyStoredAccountEnvironment(service, "XMPushService.onCreate")
        if (account != null) {
            BuildSettings.setEnvType(account.envType)
        }
        infrastructure.installMessenger()
        PushHostManagerFactory.init(service)
        val configuration = infrastructure.createConnectionConfiguration().apply {
            setDebuggerEnabled(true)
        }
        service.connectionConfiguration = configuration
        service.slimConnection = SlimConnection(service, configuration).also { connection ->
            connection.addConnectionListener(service)
        }
        service.clientEventDispatcher = service.createClientEventDispatcher()
        Alarm.initialize(service)
        service.packetSync = PacketSync(service)
        service.reconnectionManager = ReconnectionManager(service)
        CommonPacketExtensionProvider().register()
        StatsHandler.getInstance().init(service)
        service.jobController = JobScheduler("Connection Controller Thread")
        runtime.configureClientChangeListener()
        if (service.canOpenForegroundService()) {
            service.enableForegroundService()
        }
        TinyDataManager.getInstance(service).addUploader(LongConnUploader(service), TinyDataManager.UPLOADER_PUSH_CHANNEL)
        service.addPingCallBack(TinyDataCacheProcessor(service))
        service.executeJob(InitJob(service))
        service.addNetworkListener(Sync.getInstance(service))
        if (service.isPushEnabled()) {
            service.ensureConnectionChangeReceiver()
        }
        if (PushConstants.PUSH_SERVICE_PACKAGE_NAME == service.packageName) {
            infrastructure.installPowerModeObservers()
            infrastructure.installFalldownReceiver()
        }
        infrastructure.persistCreationLog(account)
    }

    fun onDestroy() {
        service.connectionChangeReceiver?.let {
            service.unregisterReceiverSafely(it)
            service.clearConnectionChangeReceiver()
        }
        service.screenStateReceiver?.let {
            service.unregisterReceiverSafely(it)
            service.clearScreenStateReceiver()
        }
        infrastructure.unregisterPowerModeObservers()
        service.clearNetworkListeners()
        service.jobController.removeAllJobs()
        service.executeJob(
            object : XMPushService.Job(XMPushServiceJob.TYPE_DISCONNECT) {
                override fun getDesc(): String = "disconnect for service destroy."

                override fun process() {
                    service.currentConnection?.disconnect(15, null)
                    service.clearCurrentConnection()
                }
            },
        )
        service.executeJob(KillJob(service))
        PushClientsManager.getInstance().removeAllClientChangeListeners()
        PushClientsManager.getInstance().resetAllClients(service, 15)
        PushClientsManager.getInstance().removeActiveClients()
        service.slimConnection.removeConnectionListener(service)
        ServiceConfig.getInstance().clear()
        Alarm.stop()
        service.clearPingCallbacks()
    }

    fun onStart(intent: Intent?, startId: Int) {
        val startedAt = System.currentTimeMillis()
        if (intent == null) {
            MyLog.e("onStart() with intent NULL")
        } else {
            MyLog.w(
                String.format(
                    "onStart() with intent.Action = %s, chid = %s, pkg = %s|%s",
                    intent.action,
                    intent.getStringExtra(PushConstants.EXTRA_CHANNEL_ID),
                    intent.getStringExtra(PushConstants.EXTRA_PACKAGE_NAME),
                    intent.getStringExtra(PushConstants.MIPUSH_EXTRA_APP_PACKAGE),
                ),
            )
        }
        intent?.action?.let { action ->
            if (PushServiceConstants.ACTION_TIMER.equals(action, ignoreCase = true) ||
                PushServiceConstants.ACTION_CHECK_ALIVE.equals(action, ignoreCase = true)
            ) {
                if (service.jobController.isBlocked()) {
                    MyLog.e("ERROR, the job controller is blocked.")
                    PushClientsManager.getInstance().resetAllClients(service, 14)
                    service.stopSelf()
                } else {
                    service.executeJob(IntentJob(service, intent))
                }
            } else if (!PushServiceConstants.ACTION_NETWORK_STATUS_CHANGED.equals(action, ignoreCase = true)) {
                service.executeJob(IntentJob(service, intent))
            }
        }
        val cost = System.currentTimeMillis() - startedAt
        if (cost > 50) {
            MyLog.v("[Prefs] spend $cost ms, too more times.")
        }
    }

    fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        onStart(intent, startId)
        return 1
    }

    fun networkChanged() {
        runtime.networkChanged()
    }

    fun connectionClosed(connection: com.xiaomi.smack.Connection, reason: Int, error: Exception?) {
        runtime.connectionClosed(connection, reason, error)
    }

    fun connectionStarted(connection: com.xiaomi.smack.Connection) {
        runtime.connectionStarted(connection)
    }

    fun postOnCreate() {
        runtime.postOnCreate()
    }

    fun reconnectionFailed(connection: com.xiaomi.smack.Connection, error: Exception) {
        runtime.reconnectionFailed(connection, error)
    }

    fun reconnectionSuccessful(connection: com.xiaomi.smack.Connection) {
        runtime.reconnectionSuccessful(connection)
    }
}
